lib/Sema/CSDiag.cpp

```diff
@@ -3659,10 +3684,33 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Re-type-check the argument with the expected type of the candidate set.
     // This should produce a specific and tailored diagnostic saying that the
     // type mismatches with expectations.
-    if (!typeCheckChildIndependently(badArgExpr,
-                                     calleeInfo.failedArgument.parameterType,
+    Type paramType = calleeInfo.failedArgument.parameterType;
+    if (!typeCheckChildIndependently(badArgExpr, paramType,
                                      CTP_CallArgument, options))
       return true;
+      
+    // If that fails, it could be that the argument doesn't conform to an archetype.
+    Type argType = badArgExpr->getType();
+
+    if (paramType->is<ArchetypeType>() && !argType->hasTypeVariable() &&
+        // FIXME: For protocol argument types, could add specific error
+        // similar to could_not_use_member_on_existential.
+        !argType->is<ProtocolType>() && !argType->is<ProtocolCompositionType>()) {
```

Comment:
@slavapestov Yes, it will fail the conformsToProtocol() check, and then this code would diagnose `argument type 'P' does not conform to expected type 'P'`, which is confusing and unhelpful. It really needs a different type of error telling the user to use a generic constraint instead.

For now the existing default error of `cannot invoke 'foo' with an argument list of type '(P)'` isn't great, but it's a little better for this situation.